### PR TITLE
Adds cache TTL and routing/caching strategy (per plugin)

### LINKED ISSUE #601: Enable per plugin cache expiration
* A plugin author should be able to expose a default cache configuration expiration value
* Users should be able to configure cache expirations through the plugin config
 * plugin config needs to be documented

### PR DESCRIPTION
**Pull request by jcooklin**
* Moves caching to the plugin pool (from plugin client)
* Moves routing strategy to the plugin pool (from availableplugins)

Resolves #601
Partially addresses #539

### CODE REVIEW

**Comment by jcooklin** [comment]
@geauxvirtual, @danielscottt, @tiffanyfj: This should be ready to go now.

File: control/available_plugin.go [hunk 1]
```diff
@@ -606,16 +657,127 @@ func (ap *availablePlugins) getPool(key string) (*apPool, serror.SnapError) {
...
+		return nil, serr
+	}
+
+	// cast client to PluginCollectorClient
+	cli, ok := p.client.(client.PluginCollectorClient)
+	if !ok {
+		return nil, serror.New(errors.New("unable to cast client to PluginCollectorClient"))
+	}
+
+	// get a metrics
```
Comment (danielscottt) [inline control/available_plugin.go]: s/a metrics/mertrics

File: control/available_plugin.go [hunk 2]
```diff
@@ -606,16 +657,127 @@ func (ap *availablePlugins) getPool(key string) (*apPool, serror.SnapError) {
...
+		results[idx] = m
+		idx++
 	}
-	return pool, nil
+	for _, m := range metricsFromCache {
+		results[idx] = m
+		idx++
+	}
+
+	// update statics about plugin
```
Comment (danielscottt) [inline control/available_plugin.go]: statistics?

File: control/available_plugin.go (hunk 1 above)
Comment (lynxbat) [inline control/available_plugin.go]: "mertrics"?

File: control/strategy/strategy.go [hunk 3]
```diff
@@ -17,17 +17,28 @@ See the License for the specific language governing permissions and
...
+	"github.com/intelsdi-x/snap/core"
+)
 
 type SelectablePlugin interface {
 	HitCount() int
 	LastHit() time.Time
 	String() string
 }
+
+type RoutingAndCaching interface {
```
Comment (danielscottt) [inline control/strategy/strategy.go]: Do you think there will be other strategy types?

I guess what I'm getting at is why not `strategy.Strategy`?

File: control/strategy/strategy.go [hunk 4]
```diff
@@ -17,17 +17,28 @@ See the License for the specific language governing permissions and
 limitations under the License.
 */
 
-package routing
+package strategy
```
Comment (danielscottt) [inline control/strategy/strategy.go]: Maybe something here to describe what package strategy is for?

File: control/available_plugin.go (hunk 1 above)
Comment (danielscottt) [inline control/available_plugin.go]: whoops.

File: control/strategy/strategy.go (hunk 3 above)
Comment (jcooklin) [inline control/strategy/strategy.go]: Originally I had strategy.Strategy and felt it was better for readability to describe what kind of strategy was being provided.  

>I guess what I'm getting at is why not strategy.Strategy?

I lean towards being more descriptive.   Refactoring the name in the future can happen when and if there is a need for another strategy type.

File: control/available_plugin.go [hunk 5]
```diff
@@ -532,24 +549,58 @@ type subscription struct {
...
-	routingStrategy RoutingStrategy
-
 	// table holds all the plugin pools.
 	// The Pools' primary keys are equal to
 	// {plugin_type}:{plugin_name}:{plugin_version}
 	table map[string]*apPool
 }
 
-func newAvailablePlugins(routingStrategy RoutingStrategy) *availablePlugins {
+type requestedMetrics struct {
```
Comment (danielscottt) [inline control/available_plugin.go]: Where are these two types used?

File: control/control.go [hunk 6]
```diff
@@ -191,7 +186,7 @@ func New(opts ...PluginControlOpt) *pluginControl {
 
 	// Plugin Runner
 	// TODO (danielscottt): handle routing strat changes via events
-	c.pluginRunner = newRunner(&routing.RoundRobinStrategy{})
```
Comment (danielscottt) [inline control/control.go]: Mind yanking this old TODO?

File: control/available_plugin.go [hunk 7]
```diff
@@ -51,12 +53,11 @@ const (
 )
 
 var (
-	// ErrPoolNotFound - error message when the plugin pool not found
```
Comment (geauxvirtual) [inline control/available_plugin.go]: Is there a reason the comments on the variables were removed?

Comment (jcooklin) [inline control/available_plugin.go]: I felt they added more clutter than value.  

@geauxvirtual: do you feel they should be put back?

**Comment by danielscottt** [comment]
LGTM

**Pull request merged**